File: 0400_0599/leetcode-no594.cpp
#include <iostream>
#include <map>
#include <vector>

using namespace std;

class Solution {
private:
    map<int, int> cnt;
public:
    int findLHS(vector<int> &nums) {
        for (auto i: nums) {
            cnt[i]++;
        }
        vector<pair<int,int>> temp;
        for (auto p : cnt) {
            temp.push_back(p);
        }
        int maxium = 0;
        for (int i = 1;i < temp.size();i++){
            if (temp[i].first - temp[i - 1].first == 1){
                maxium = max(maxium,temp[i].second + temp[i - 1].second);
            }
        }
        return maxium;
    }
};

int main() {
    Solution test;
    vector<int> nums = {1,3,2,2,5,2,3,7};
    cout << test.findLHS(nums) << endl;
}
//
// Created by ASUS on 2024/2/21.
//

File: leetcode-rs/0500_0599/leetcode-no594.rs
// problem: https://leetcode.cn/problems/longest-harmonious-subsequence/description/
struct Solution;

use std::cmp::max;
use std::collections::HashMap;
impl Solution {
    pub fn find_lhs(nums: Vec<i32>) -> i32 {
        let mut cnts: HashMap<i32, i32> = HashMap::new();
        for num in nums {
            *cnts.entry(num).or_default() += 1;
        }
        let mut res: i32 = 0;
        for num in &cnts {
            if let Some(T) = cnts.get(&(num.0 + 1)) {
                res = max(res, num.1 + T);
            }
        }
        res
    }
}

fn main() {
    println!("{:?}", Solution::find_lhs(vec![1, 3, 2, 2, 5, 2, 3, 7]));
    println!("{:?}", Solution::find_lhs(vec![1,1,1,1]));
}

//
// Created by ASUS at 6/30/2025 10:42 PM
//